### ISSUE TITLE
"run_shebang" sometimes fails with "Text file busy"

### ISSUE THREAD

**Issue by 06kellyjac** [casey/just#855]
Moved here from #849 

> > Is this a chooser test? https://github.com/casey/just/blob/50cd24d37ba696dd74f3b05f1c23b934ef049b10/src/justfile.rs#L429-L450
> >
> >I've had it fail randomly a couple times with "Text file busy"
>
> @06kellyjac That's interesting, that's not a chooser test. Is that the only test you've seen fail with that error?

Yeah and I couldn't get to fail consistently. It failed when building the latest update using nix with this patching (the original sed on the left): https://github.com/NixOS/nixpkgs/pull/125653/files#diff-85969503e7886ff0bcdf36e5c35be3b9fa6cd334be07e99d9948a4643bcbe90eL38-L43

We patch these shebangs because `env` doesn't have access to anything in the sandbox.
I thought there might be an issue with the patching of `#!/usr/bin/env sh` to use bash, which could be used at the same time as another test that uses `bash` but I don't see why there'd be a "Text file busy" error for two processes trying to run `bash` :shrug:

I don't have the output of the issue in my terminal anymore so I'll try recreate the error
I'll also move this discussion to a new issue to avoid filling this one with non-chooser test stuff

**Comment by 06kellyjac** [comment]
The only thing that should be running in the sandbox should be the just tests.
Maybe I cleaned up some dependencies using `nix-collect-garbage` but that *should* leave the dependencies currently in use (which should include the bash version used mid test)

Also the copy of bash is readonly

```
λ ls -l $(readlink -f $(which bash))
.r-xr-xr-x root root 988 KB Thu Jan  1 01:00:01 1970  /nix/store/fgbzvd4c6nly9m4dpczrxybdpkm8mnk3-bash-interactive-4.4-p23/bin/bash
```

**Comment by 06kellyjac** [comment]
Hmm, I've done 5 builds from scratch using the setup I believe ran into the error before but now it just wont happen...

**Comment by casey** [comment]
A Heisenbug! Let's leave this issue open. I think I've seen it happen, and perhaps someone else will be able to reproduce it in the future.

**Comment by 06kellyjac** [comment]
Got one :slightly_smiling_face: 

```
++ cargo test -j 12 --release --target x86_64-unknown-linux-gnu --frozen -- --test-threads=12
   Compiling cfg-if v1.0.0
# ...
   Compiling env_logger v0.8.3
   Compiling tempfile v3.2.0
   Compiling test-utilities v0.0.0 (/build/source/test-utilities)
   Compiling ctrlc v3.1.9
   Compiling just v0.9.4 (/build/source)
    Finished release [optimized] target(s) in 1m 24s
     Running unittests (target/x86_64-unknown-linux-gnu/release/deps/just-1a802a2867333503)

running 430 tests
test analyzer::tests::duplicate_recipe ... ok
# ...
test unindent::tests::unindents ... ok

failures:

---- justfile::tests::run_shebang stdout ----
thread 'justfile::tests::run_shebang' panicked at 'Unexpected run error: Shebang { recipe: "a", command: "/nix/store/a4yw1svqqk4d8lhwinn9xp847zz9gfma-bash-4.4-p23/bin/sh", argument: None, io_error: Os { code: 26, kind: Other, message: "Text file busy" } }', src/justfile.rs:429:3
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    justfile::tests::run_shebang

test result: FAILED. 429 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s

error: test failed, to rerun pass '--lib'
error: builder for '/nix/store/jk03yiqzf2afcr2g96bzwfgy74l8fcd1-just-0.9.4.drv' failed with exit code 101;
       last 10 log lines:
       > thread 'justfile::tests::run_shebang' panicked at 'Unexpected run error: Shebang { recipe: "a", command: "/nix/store/a4yw1svqqk4d8lhwinn9xp847zz9gfma-bash-4.4-p23/bin/sh", argument: None, io_error: Os { code: 26, kind: Other, message: "Text file busy" } }', src/justfile.rs:429:3
       > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
       >
       >
       > failures:
       >     justfile::tests::run_shebang
       >
       > test result: FAILED. 429 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s
       >
       > error: test failed, to rerun pass '--lib'
       For full logs, run 'nix log /nix/store/jk03yiqzf2afcr2g96bzwfgy74l8fcd1-just-0.9.4.drv'.
```

This time I was 100% not messing around with `sh`/`bash`, the file definitely exists and is read only, and I haven't seen any commands that should touch the `just` binary either

```
λ ls -al /nix/store/a4yw1svqqk4d8lhwinn9xp847zz9gfma-bash-4.4-p23/bin/sh
lrwxrwxrwx root root 4 B Thu Jan  1 01:00:01 1970  /nix/store/a4yw1svqqk4d8lhwinn9xp847zz9gfma-bash-4.4-p23/bin/sh ⇒ bash
```

Full log: https://gist.github.com/06kellyjac/e23d799ab4561f9034f04769e1b72b78

**Comment by casey** [comment]
I'm mystified. As far as I know, `just` isn't doing anything that jumps out at me that would provoke this.

In the error:

```rust
Shebang {
  recipe: "a",
  command: "/nix/store/a4yw1svqqk4d8lhwinn9xp847zz9gfma-bash-4.4-p23/bin/sh",
  argument: None,
  io_error: Os { code: 26, kind: Other, message: "Text file busy" },
}
```

`argument` is `None`, but `just` is invoking a shebang recipe (something where the first line is `#!`, so the real command is something like

```
/nix/store/a4yw1svqqk4d8lhwinn9xp847zz9gfma-bash-4.4-p23/bin/sh path-to-tempfile
```

Where `path-to-tempfile` is a file that contains the extracted recipe text. In this case it's the recipe from the `run_shebang` test, so the textfile will contain this:

```
#!/usr/bin/env sh
code=200
x() { return $code; }
  x
    x
```

(I'm not really sure why the formatting is so weird.)

Is is possible that the error message regards the tempfile?

**Comment by cinerea0** [comment]
Just encountered this while building using xbps-src for Void Linux:

```
failures:

---- justfile::tests::run_shebang stdout ----
thread 'justfile::tests::run_shebang' panicked at src/justfile.rs:585:3:
Unexpected run error: Shebang { recipe: "a", command: "/usr/bin/env", argument: Some("sh"), io_error: Os { code: 26, kind: ExecutableFileBusy, message: "Text file busy" } }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    justfile::tests::run_shebang

test result: FAILED. 469 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

error: test failed, to rerun pass `--lib`
```

After I re-ran the build it succeeded. Very strange!

**Comment by casey** [comment]
Yeah this one is extremely curious. It's intermittent, and I don't think anyone has been able to come up with a consistent reproduction.

**Comment by yonas** [comment]
Found in Debian packaging CI: https://ci.debian.net/packages/r/rust-just/testing/amd64/41726210/#L5640

**Comment by Skgland** [comment]
```sh
#!/bin/bash

export RUST_BACKTRACE=full

run()
{
    cargo test --offline --  justfile::
}

cargo fetch

while run ; do
:
done
```
is consistently reproducing this for me in less than a minute.

**Comment by casey** [comment]
@Skgland thanks for the repro! I was able to reproduce this on Linux, but not on macOS.

I just merged #1880, which I think fixes this. I'm still not sure what the underlying issue is, but `run_shebang` was one of a few unit tests which actually run a recipe in-process. I'm guessing that there's some kind of race condition that involves running the shebang recipe script while another process is modifying it. I converted run_shebang into an integration test which calls the `just` binary, instead of running in-process, and it seems to have fixed it.
